File: Between/Between/Services/ScheduleEngine.swift
import Foundation

enum ScheduleEngine {
    /// Campus day window — avoids 6am–11pm empty blocks in the UI.
    static let startMinutes = 8 * 60
    static let endMinutes = 18 * 60
    static let minFreeBlockMinutes = 30
    static let minOverlapMinutes = 25
    /// Hide empty free blocks longer than this; overlaps always show.
    static let maxEmptyFreeBlockMinutes = 90

    private enum TimelineBlock {
        case free(start: Int, end: Int)
        case busy(start: Int, end: Int, section: CourseSection)
    }

    private static let dayMap: [String: Int] = [
        "Sun": 0, "Mon": 1, "Tue": 2, "Wed": 3, "Thu": 4, "Fri": 5, "Sat": 6
    ]

    static func formatRange(start: Int, end: Int) -> String {
        return "\(formatTime12Hour(start)) – \(formatTime12Hour(end))"
    }

    static func formatDuration(_ minutes: Int) -> String {
        if minutes < 60 { return "\(minutes) min" }
        let h = minutes / 60
        let m = minutes % 60
        if m == 0 { return h == 1 ? "1 hr" : "\(h) hr" }
        return "\(h) hr \(m) min"
    }

    static func buildTodayPlan(
        mySections: [CourseSection],
        friendSectionsById: [String: [CourseSection]],
        friendNamesById: [String: String],
        now: Date = Date()
    ) -> [TodayPlanItem] {
        let dayIdx = BackendConfiguration.demoWeekdayIndex ?? todayIndex(from: now)
        let nowMinutes = BackendConfiguration.demoNowMinutes
            ?? (Calendar.current.component(.hour, from: now) * 60
                + Calendar.current.component(.minute, from: now))

        let busy = busyIntervals(on: dayIdx, sections: mySections)
        let free = freeIntervals(on: dayIdx, sections: mySections)

        var timeline: [TodayPlanItem] = []

        var freeIndex = 0
        var busyIndex = 0

        while freeIndex < free.count || busyIndex < busy.count {
            let nextFree = freeIndex < free.count ? free[freeIndex] : nil
            let nextBusy = busyIndex < busy.count ? busy[busyIndex] : nil

            guard let block = nextTimelineBlock(free: nextFree, busy: nextBusy) else { break }

            switch block {
            case let .free(start, end):
                let overlaps = friendOverlaps(
                    start: start, end: end, dayIdx: dayIdx,
                    friendSectionsById: friendSectionsById,
                    friendNamesById: friendNamesById
                )
                if overlaps.isEmpty {
                    let duration = end - start
                    if duration >= minFreeBlockMinutes && duration <= maxEmptyFreeBlockMinutes {
                        appendIfFuture(
                            TodayPlanItem(
                                id: "free-\(start)-\(end)",
                                kind: .freeBlock,
                                startMinutes: start,
                                endMinutes: end,
                                section: nil,
                                friendOverlaps: []
                            ),
                            nowMinutes: nowMinutes,
                            into: &timeline
                        )
                    }
                } else {
                    for overlap in overlaps {
                        guard let best = overlap.intervals
                            .filter({ $0.end - $0.start >= minOverlapMinutes })
                            .max(by: { ($0.end - $0.start) < ($1.end - $1.start) })
                        else { continue }

                        let clipped = FriendOverlap(
                            id: "\(overlap.friendId)-\(best.start)",
                            friendId: overlap.friendId,
                            friendName: overlap.friendName,
                            intervals: [best],
                            totalMinutes: best.end - best.start
                        )
                        appendIfFuture(
                            TodayPlanItem(
                                id: "overlap-\(overlap.friendId)-\(best.start)",
                                kind: .freeBlock,
                                startMinutes: best.start,
                                endMinutes: best.end,
                                section: nil,
                                friendOverlaps: [clipped]
                            ),
                            nowMinutes: nowMinutes,
                            into: &timeline
                        )
                    }
                }
                freeIndex += 1

            case let .busy(start, end, section):
                appendIfFuture(
                    TodayPlanItem(
                        id: "class-\(section.sectionId)",
                        kind: .classBlock,
                        startMinutes: start,
                        endMinutes: end,
                        section: section,
                        friendOverlaps: []
                    ),
                    nowMinutes: nowMinutes,
                    into: &timeline
                )
                busyIndex += 1
            }
        }

        return timeline.sorted { $0.startMinutes < $1.startMinutes }
    }

    // MARK: - Private helpers

    private static func todayIndex(from date: Date) -> Int {
        Calendar.current.component(.weekday, from: date) - 1
    }

    private static func minutes(from time: String) -> Int? {
        let parts = time.split(separator: ":")
        guard parts.count == 2, let hour = Int(parts[0]), let minute = Int(parts[1]) else { return nil }
        return hour * 60 + minute
    }

    static func formatTime12Hour(_ minutes: Int) -> String {
        let hours = minutes / 60
        let mins = minutes % 60
        let h12 = hours == 0 ? 12 : (hours > 12 ? hours - 12 : hours)
        let ampm = hours < 12 ? "AM" : "PM"
        if mins == 0 { return "\(h12) \(ampm)" }
        return String(format: "%d:%02d %@", h12, mins, ampm)
    }

    private static func sectionsForDay(_ dayIdx: Int, in allSections: [CourseSection]) -> [CourseSection] {
        allSections.filter { section in
            section.meetingDays.contains { dayMap[$0] == dayIdx }
        }
    }

    private static func busyIntervals(
        on dayIdx: Int,
        sections: [CourseSection]
    ) -> [(start: Int, end: Int, section: CourseSection)] {
        sectionsForDay(dayIdx, in: sections).compactMap { section in
            guard let start = minutes(from: section.startTime),
                  let end = minutes(from: section.endTime) else { return nil }
            let clampedStart = max(start, startMinutes)
            let clampedEnd = min(end, endMinutes)
            guard clampedEnd > clampedStart else { return nil }
            return (start: clampedStart, end: clampedEnd, section: section)
        }
        .sorted { $0.start < $1.start }
    }

    private static func freeIntervals(
        on dayIdx: Int,
        sections: [CourseSection]
    ) -> [(start: Int, end: Int)] {
        let busy = busyIntervals(on: dayIdx, sections: sections).map { (start: $0.start, end: $0.end) }
        var free: [(start: Int, end: Int)] = []
        var previousEnd = startMinutes
        for interval in busy {
            if interval.start > previousEnd {
                free.append((start: previousEnd, end: interval.start))
            }
            previousEnd = max(previousEnd, interval.end)
        }
        if previousEnd < endMinutes {
            free.append((start: previousEnd, end: endMinutes))
        }
        return free
    }

    private static func intersectIntervals(
        _ a: [(start: Int, end: Int)],
        _ b: [(start: Int, end: Int)]
    ) -> [(start: Int, end: Int)] {
        var result: [(start: Int, end: Int)] = []
        let sortedA = a.sorted { $0.start < $1.start }
        let sortedB = b.sorted { $0.start < $1.start }
        var i = 0
        var j = 0
        while i < sortedA.count && j < sortedB.count {
            let left = sortedA[i]
            let right = sortedB[j]
            let start = max(left.start, right.start)
            let end = min(left.end, right.end)
            if end > start { result.append((start: start, end: end)) }
            if left.end < right.end { i += 1 } else { j += 1 }
        }
        return result
    }

    private static func nextTimelineBlock(
        free: (start: Int, end: Int)?,
        busy: (start: Int, end: Int, section: CourseSection)?
    ) -> TimelineBlock? {
        switch (free, busy) {
        case let (freeBlock?, nil): return .free(start: freeBlock.start, end: freeBlock.end)
        case let (nil, classBlock?): return .busy(start: classBlock.start, end: classBlock.end, section: classBlock.section)
        case let (freeBlock?, classBlock?):
            return freeBlock.start <= classBlock.start
                ? .free(start: freeBlock.start, end: freeBlock.end)
                : .busy(start: classBlock.start, end: classBlock.end, section: classBlock.section)
        case (nil, nil): return nil
        }
    }

    private static func appendIfFuture(
        _ item: TodayPlanItem,
        nowMinutes: Int,
        into timeline: inout [TodayPlanItem]
    ) {
        guard item.endMinutes > nowMinutes else { return }
        if item.startMinutes < nowMinutes {
            timeline.append(
                TodayPlanItem(
                    id: item.id,
                    kind: item.kind,
                    startMinutes: nowMinutes,
                    endMinutes: item.endMinutes,
                    section: item.section,
                    friendOverlaps: clipOverlaps(item.friendOverlaps, nowMinutes: nowMinutes)
                )
            )
        } else {
            timeline.append(item)
        }
    }

    private static func clipOverlaps(_ overlaps: [FriendOverlap], nowMinutes: Int) -> [FriendOverlap] {
        overlaps.compactMap { overlap in
            let intervals = overlap.intervals.compactMap { interval -> (start: Int, end: Int)? in
                let start = max(interval.start, nowMinutes)
                guard interval.end > start, interval.end - start >= minOverlapMinutes else { return nil }
                return (start: start, end: interval.end)
            }
            guard !intervals.isEmpty else { return nil }
            let total = intervals.reduce(0) { $0 + ($1.end - $1.start) }
            return FriendOverlap(
                id: overlap.id,
                friendId: overlap.friendId,
                friendName: overlap.friendName,
                intervals: intervals,
                totalMinutes: total
            )
        }
    }

    private static func friendOverlaps(
        start: Int,
        end: Int,
        dayIdx: Int,
        friendSectionsById: [String: [CourseSection]],
        friendNamesById: [String: String]
    ) -> [FriendOverlap] {
        let userFree = [(start: start, end: end)]
        var overlaps: [FriendOverlap] = []

        for (friendId, sections) in friendSectionsById {
            let friendFree = freeIntervals(on: dayIdx, sections: sections)
            let intervals = intersectIntervals(userFree, friendFree)
            let qualifying = intervals.filter { $0.end - $0.start >= minOverlapMinutes }
            guard !qualifying.isEmpty else { continue }
            let totalMinutes = qualifying.reduce(0) { $0 + ($1.end - $1.start) }
            overlaps.append(
                FriendOverlap(
                    id: "\(friendId)-\(start)",
                    friendId: friendId,
                    friendName: friendNamesById[friendId] ?? "Friend",
                    intervals: qualifying,
                    totalMinutes: totalMinutes
                )
            )
        }

        return overlaps.sorted {
            if $0.totalMinutes != $1.totalMinutes { return $0.totalMinutes > $1.totalMinutes }
            return $0.friendName < $1.friendName
        }
    }
}

File: Between/api/v1/scheduleEngine.js
/**
 * Campus schedule overlap engine — mirrors iOS ScheduleEngine.swift
 */

const DAY_MAP = { Sun: 0, Mon: 1, Tue: 2, Wed: 3, Thu: 4, Fri: 5, Sat: 6 };

const START_MINUTES = 8 * 60;
const END_MINUTES = 18 * 60;
const MIN_FREE_BLOCK = 30;
const MIN_OVERLAP = 25;
const MAX_EMPTY_FREE = 90;

const DEMO_WEEKDAY = 3; // Wednesday
const DEMO_NOW = 10 * 60 + 15;

function minutesFromTime(time) {
  const [h, m] = time.split(':').map(Number);
  return h * 60 + m;
}

function formatTime12Hour(minutes) {
  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;
  const h12 = hours === 0 ? 12 : hours > 12 ? hours - 12 : hours;
  const ampm = hours < 12 ? 'AM' : 'PM';
  if (mins === 0) return `${h12} ${ampm}`;
  return `${h12}:${String(mins).padStart(2, '0')} ${ampm}`;
}

function formatRange(start, end) {
  return `${formatTime12Hour(start)} – ${formatTime12Hour(end)}`;
}

function sectionsForDay(dayIdx, sections) {
  return sections.filter((s) => s.meetingDays.some((d) => DAY_MAP[d] === dayIdx));
}

function busyIntervals(dayIdx, sections) {
  return sectionsForDay(dayIdx, sections)
    .map((section) => {
      const start = minutesFromTime(section.startTime);
      const end = minutesFromTime(section.endTime);
      const clampedStart = Math.max(start, START_MINUTES);
      const clampedEnd = Math.min(end, END_MINUTES);
      if (clampedEnd <= clampedStart) return null;
      return { start: clampedStart, end: clampedEnd, section };
    })
    .filter(Boolean)
    .sort((a, b) => a.start - b.start);
}

function freeIntervals(dayIdx, sections) {
  const busy = busyIntervals(dayIdx, sections).map(({ start, end }) => ({ start, end }));
  const free = [];
  let prev = START_MINUTES;
  for (const b of busy) {
    if (b.start > prev) free.push({ start: prev, end: b.start });
    prev = Math.max(prev, b.end);
  }
  if (prev < END_MINUTES) free.push({ start: prev, end: END_MINUTES });
  return free;
}

function intersectIntervals(a, b) {
  const result = [];
  const sa = [...a].sort((x, y) => x.start - y.start);
  const sb = [...b].sort((x, y) => x.start - y.start);
  let i = 0;
  let j = 0;
  while (i < sa.length && j < sb.length) {
    const start = Math.max(sa[i].start, sb[j].start);
    const end = Math.min(sa[i].end, sb[j].end);
    if (end > start) result.push({ start, end });
    if (sa[i].end < sb[j].end) i++;
    else j++;
  }
  return result;
}

function clipOverlaps(overlaps, nowMinutes) {
  return overlaps
    .map((o) => {
      const intervals = o.intervals
        .map(({ start, end }) => {
          const s = Math.max(start, nowMinutes);
          return end > s && end - s >= MIN_OVERLAP ? { start: s, end } : null;
        })
        .filter(Boolean);
      if (!intervals.length) return null;
      const total = intervals.reduce((sum, iv) => sum + (iv.end - iv.start), 0);
      return { ...o, intervals, totalMinutes: total };
    })
    .filter(Boolean);
}

function appendIfFuture(item, nowMinutes, timeline) {
  if (item.endMinutes <= nowMinutes) return;
  if (item.startMinutes < nowMinutes) {
    timeline.push({
      ...item,
      startMinutes: nowMinutes,
      friendOverlaps: clipOverlaps(item.friendOverlaps, nowMinutes),
    });
  } else {
    timeline.push(item);
  }
}

function friendOverlaps(start, end, dayIdx, friendSectionsById, friendNamesById) {
  const userFree = [{ start, end }];
  const overlaps = [];
  for (const [friendId, sections] of Object.entries(friendSectionsById)) {
    const friendFree = freeIntervals(dayIdx, sections);
    const intervals = intersectIntervals(userFree, friendFree).filter(
      (iv) => iv.end - iv.start >= MIN_OVERLAP
    );
    if (!intervals.length) continue;
    const totalMinutes = intervals.reduce((s, iv) => s + (iv.end - iv.start), 0);
    overlaps.push({
      id: `${friendId}-${start}`,
      friendId,
      friendName: friendNamesById[friendId] || 'Friend',
      intervals,
      totalMinutes,
    });
  }
  return overlaps.sort((a, b) => b.totalMinutes - a.totalMinutes || a.friendName.localeCompare(b.friendName));
}

function buildTodayPlan(mySections, friendSectionsById, friendNamesById) {
  const dayIdx = DEMO_WEEKDAY;
  const nowMinutes = DEMO_NOW;
  const free = freeIntervals(dayIdx, mySections);
  const busy = busyIntervals(dayIdx, mySections);
  const timeline = [];
  let fi = 0;
  let bi = 0;

  while (fi < free.length || bi < busy.length) {
    const nf = fi < free.length ? free[fi] : null;
    const nb = bi < busy.length ? busy[bi] : null;
    if (!nf && !nb) break;

    if (nf && (!nb || nf.start <= nb.start)) {
      const overlaps = friendOverlaps(nf.start, nf.end, dayIdx, friendSectionsById, friendNamesById);
      if (!overlaps.length) {
        const duration = nf.end - nf.start;
        if (duration >= MIN_FREE_BLOCK && duration <= MAX_EMPTY_FREE) {
          appendIfFuture(
            {
              id: `free-${nf.start}-${nf.end}`,
              kind: 'freeBlock',
              startMinutes: nf.start,
              endMinutes: nf.end,
              section: null,
              friendOverlaps: [],
            },
            nowMinutes,
            timeline
          );
        }
      } else {
        for (const overlap of overlaps) {
          const best = overlap.intervals
            .filter((iv) => iv.end - iv.start >= MIN_OVERLAP)
            .sort((a, b) => b.end - b.start - (a.end - a.start))[0];
          if (!best) continue;
          appendIfFuture(
            {
              id: `overlap-${overlap.friendId}-${best.start}`,
              kind: 'freeBlock',
              startMinutes: best.start,
              endMinutes: best.end,
              section: null,
              friendOverlaps: [
                {
                  id: `${overlap.friendId}-${best.start}`,
                  friendId: overlap.friendId,
                  friendName: overlap.friendName,
                  intervals: [best],
                  totalMinutes: best.end - best.start,
                },
              ],
            },
            nowMinutes,
            timeline
          );
        }
      }
      fi++;
    } else if (nb) {
      appendIfFuture(
        {
          id: `class-${nb.section.sectionId}`,
          kind: 'classBlock',
          startMinutes: nb.start,
          endMinutes: nb.end,
          section: nb.section,
          friendOverlaps: [],
        },
        nowMinutes,
        timeline
      );
      bi++;
    }
  }

  return timeline.sort((a, b) => a.startMinutes - b.startMinutes);
}

function serializeTodayPlan(items) {
  return items.map((item) => ({
    id: item.id,
    kind: item.kind,
    startMinutes: item.startMinutes,
    endMinutes: item.endMinutes,
    section: item.section,
    friendOverlaps: item.friendOverlaps.map((o) => ({
      id: o.id,
      friendId: o.friendId,
      friendName: o.friendName,
      intervals: o.intervals.map((iv) => [iv.start, iv.end]),
      totalMinutes: o.totalMinutes,
    })),
  }));
}

module.exports = {
  buildTodayPlan,
  serializeTodayPlan,
  formatRange,
  formatTime12Hour,
  DEMO_WEEKDAY,
};
